rest_specification/restful_service_specification.md

```diff
@@ -0,0 +1,240 @@
+performance or other considerations; for example, `GET` may not be supported
+for `user` if listing all users is disallowed for performance reasons.
+
+## Operations Against Specific Resources
+Specific resources will be addressed through the base URI plus the identifier
+for that resource.
+
+    /user/{id}
+
+The following operations may be supported:
+
+| HTTP Operation | Description                   | Query Parameters | Request Body | Response Body |
+|----------------|-------------------------------|------------------|--------------|---------------|
+| `GET`          | Retrieve the resource         | `fields`         | -            | JSON object   |
+| `PUT`          | Replace the existing resource | -                | JSON object  | -             |
+| `PATCH`        | Update an existing resource   | -                | JSON object  | -             |
+| `DELETE`       | Delete the resource           | -                | -            | -             |
+
+Clients should use the `fields` parameter for better performance.
+
+Implementors should consider breaking out field represented by a JSON array
+into sub-entities if there are performance concerns around its size. This
+consideration should happen regardless of how the data is stored.
+
+## Pagination
+
+Paging shall be implemented via `page` and `size` query parameters. If none are
+provided, the default page size shall be used.
+
+The default page-size is implementation-specific.
+
+| Parameter | Description               |
+|-----------|---------------------------|
+| `page`    | page index, starting at 0 |
+| `size`    | page size                 |
+
+## Sorting
+
+Sorting shall be provided via the `sort` query parameter. `sort` may be
+specified more than once in order to sort by more than one field. `sort` values
+shall be of the form `sort={field}(,{direction})` where `direction` is `asc` or
+`desc`.
+
+### Responses
+
+The response shall be a JSON object formatted according to Spring Data REST
+conventions.Specifically, the list is returned as a pluralized field name
+under `_embedded`, and a `page` object supplies paging info. Example:
+
+```json
+GET /user
+
+{
+  "_embedded": {
+    "users": [
+      {
+        ...
+      },
+      ...
+    ]
+  },
+  "page": {
+    "size": 20,
+    "totalElements": 200,
+    "totalPages": 10,
+    "number": 0
+  }
+}
+```
+
+## Additional Functions
+
+Additional APIs may expose other functionality related to a resource type.
+Examples include:
+ - query for a specific subset of resources
+ - perform an action against a set of resources
+
+Additional functions shall be `POST` only, and parameters will be provided in
+the request body. The request body SHOULD be a JSON object with fields
+corresponding to parameter names.
+Examples:
+
+```json
+POST /user/findByLogin
+
+{
+  "login": "bob"
+}
+```
+
+Each function that can be performed against a specific resource must exist in
+two forms:
+ - non-batch
+ - batch
```

Comment:
That's changing: we are adding the concept of 'ordered' and 'unordered'
bulk requests. Unordered bulk requests will execute in parallel with no
particular order. Ordered bulk requests are processed in order, with only
consecutive find requests running in parallel.

On Fri, Jun 10, 2016 at 8:02 AM, Dennis Crissman <notifications@github.com>
wrote:

> In rest_specification/restful_service_specification.md
> <https://github.com/lightblue-platform/lightblue-docs-specs/pull/43#discussion_r66617611>
> :
>
> > +the request body. The request body SHOULD be a JSON object with fields
> > +corresponding to parameter names.
> > +Examples:
> > +
> > +```json
> > +POST /user/findByLogin
> > +
> > +{
> > +  "login": "bob"
> > +}
> > +```
> > +
> > +Each function that can be performed against a specific resource must exist in
> > +two forms:
> > + - non-batch
> > + - batch
>
> I am curious if we should support async operations for bulk requests.
>
> From a lightblue perspective, only find can be performed async. All other
> operations are always sync in order to ensure order of operation.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/lightblue-platform/lightblue-docs-specs/pull/43/files/19a07e2017455997f1b9c4bdb15edb6e3cceb8ce#r66617611>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ADgDDd8mAhU3Ls3YU53hwFBF73MjJk4Jks5qKW5xgaJpZM4IopeE>
> .
>